common/src/main/java/io/netty/util/NetUtil.java

```diff
@@ -439,164 +423,148 @@ public static String bytesToIpAddress(byte[] bytes, int offset, int length) {
+                wordStart = curIdx + 1;
+                wordLen = 0;
+                break;
+            case '%':
+                // Strip the interface name/index after the percent sign
+                break loop;
+            default:
+                if (wordLen > 3 || !isValidHexChar(c)) {
+                    return false;
+                }
+                wordLen++;
+                prevColon = false;
             }
         }
 
         // Check if we have an IPv4 ending
-        if (numberOfPeriods > 0) {
-            // There is a test case with 7 colons and valid ipv4 this should resolve it
-            if (numberOfPeriods != 3 || !(isValidIp4Word(word.toString()) && (numberOfColons < 7 || doubleColon))) {
-                return false;
-            }
-        } else {
-            // If we're at then end and we haven't had 7 colons then there is a
-            // problem unless we encountered a doubleColon
-            if (numberOfColons != 7 && !doubleColon) {
-                return false;
-            }
-
-            // If we have an empty word at the end, it means we ended in either
-            // a : or a .
-            // If we did not end in :: then this is invalid
-            if (word.length() == 0 && ipAddress.charAt(length - 1 - startOffset) == ':' &&
-                ipAddress.charAt(length - 2 - startOffset) != ':') {
-                return false;
-            }
+        if (dots > 0) {
+            return dots == 3 && isValidIpV4Word(ip, wordStart, wordStart + wordLen);
         }
 
-        return true;
+        // If we're at then end and we haven't had 7 colons then there is a
+        // problem unless we encountered a doubleColon
+        return (colons == 7 || doubleColon) &&
+               // If we have an empty word at the end, it means we ended in either a : or a .
+               // If we did not end in :: then this is invalid
+               (wordLen > 0 || (prevColon && doubleColon));
     }
 
-    private static boolean isValidIp4Word(String word) {
-        char c;
-        if (word.length() < 1 || word.length() > 3) {
+    private static boolean isValidIpV4Word(CharSequence word, int start, int endExclusive) {
+        int len = endExclusive - start;
+        char c0, c1, c2;
+        if (len < 1 || len > 3 || (c0 = word.charAt(start)) < '0') {
             return false;
         }
-        for (int i = 0; i < word.length(); i ++) {
-            c = word.charAt(i);
-            if (!(c >= '0' && c <= '9')) {
-                return false;
-            }
+        if (len == 3) {
+            return (c1 = word.charAt(start + 1)) >= '0' &&
+                   (c2 = word.charAt(start + 2)) >= '0' &&
+                   (c0 <= '1' && c1 <= '9' && c2 <= '9' ||
+                    c0 == '2' && c1 <= '5' && (c2 <= '5' || c1 < '5' && c2 <= '9'));
         }
-        return Integer.parseInt(word) <= 255;
+        return c0 <= '9' && (len == 1 || isValidNumericChar(word.charAt(start + 1)));
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Reduce the number of conditional operators (9) used in the expression (maximum allowed 5). [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1067)